File: .skills/openclaw-skills/skills/0xnyk/xint/lib/top.ts
import { loadTokens, getValidToken } from "./oauth";
import { getUserTimeline, type TweetWithEngagement } from "./timeline";
import { bold, cyan, yellow, dim, green } from "./format";

type SortMetric = "likes" | "impressions" | "clicks" | "engagement_rate";
type ContentType = "thread" | "media" | "single";

function classifyTweet(t: TweetWithEngagement): ContentType {
  if (t.conversation_id === t.id) return "thread";
  const hasMedia = (t.urls || []).some(u => u.url?.includes("pic.twitter.com")) ||
                   t.text.includes("pic.twitter.com");
  if (hasMedia) return "media";
  return "single";
}

function getSortValue(t: TweetWithEngagement, metric: SortMetric): number {
  switch (metric) {
    case "likes": return t.metrics.likes;
    case "impressions": return t.metrics.impressions;
    case "clicks": return t.non_public_metrics?.url_link_clicks || 0;
    case "engagement_rate": return t.engagement_rate;
  }
}

function fmtPct(n: number): string {
  return (n * 100).toFixed(2) + "%";
}

function fmtNum(n: number): string {
  if (n >= 1_000_000) return (n / 1_000_000).toFixed(1) + "M";
  if (n >= 1_000) return (n / 1_000).toFixed(1) + "K";
  return String(n);
}

export async function cmdTop(args: string[]): Promise<void> {
  let sortBy: SortMetric = "engagement_rate";
  let limit = 10;
  let since = "7d";
  let filterType: ContentType | undefined;
  let asJson = false;
  let pages = 3;

  let i = 0;
  while (i < args.length) {
    const arg = args[i];
    switch (arg) {
      case "--by":
        sortBy = (args[++i] || "engagement_rate") as SortMetric;
        break;
      case "--limit":
        limit = parseInt(args[++i] || "10");
        break;
      case "--since":
        since = args[++i] || "7d";
        break;
      case "--type":
        filterType = (args[++i] || undefined) as ContentType | undefined;
        break;
      case "--json":
        asJson = true;
        break;
      case "--pages":
        pages = Math.min(parseInt(args[++i] || "3"), 10);
        break;
      case "--help":
      case "-h":
        printHelp();
        return;
      default:
        break;
    }
    i++;
  }

  let token: string;
  try {
    token = await getValidToken();
  } catch (e: any) {
    console.error("OAuth required. Run 'xint auth setup' first.");
    console.error(`Error: ${e.message}`);
    process.exit(1);
    return;
  }

  const tokens = loadTokens();
  if (!tokens?.user_id) {
    console.error("Could not determine user ID from stored tokens. Re-run 'xint auth setup'.");
    process.exit(1);
    return;
  }

  console.error(`Fetching your tweets (since ${since})...`);

  let tweets = await getUserTimeline(tokens.user_id, token, {
    since,
    pages,
    exclude: ["replies", "retweets"],
  });

  if (filterType) {
    tweets = tweets.filter(t => classifyTweet(t) === filterType);
  }

  tweets.sort((a, b) => getSortValue(b, sortBy) - getSortValue(a, sortBy));
  const top = tweets.slice(0, limit);

  if (top.length === 0) {
    console.error("No tweets found in the specified period.");
    return;
  }

  console.error(`Found ${tweets.length} tweets, showing top ${top.length} by ${sortBy}\n`);

  if (asJson) {
    console.log(JSON.stringify(top.map(t => ({
      id: t.id,
      text: t.text,
      type: classifyTweet(t),
      engagement_rate: t.engagement_rate,
      metrics: t.metrics,
      tweet_url: t.tweet_url,
      created_at: t.created_at,
    })), null, 2));
    return;
  }

  console.log(bold(`Top ${top.length} Tweets by ${sortBy}`) + "\n");
  for (let j = 0; j < top.length; j++) {
    const t = top[j];
    const preview = t.text.replace(/https:\/\/t\.co\/\S+/g, "").trim().slice(0, 100);
    const type = classifyTweet(t);
    const typeLabel = type === "thread" ? "[thread]" : type === "media" ? "[media]" : "[single]";

    console.log(`${cyan(String(j + 1))}. ${dim(typeLabel)} ${yellow(fmtPct(t.engagement_rate))} engagement`);
    console.log(`   ${preview}${preview.length >= 100 ? "..." : ""}`);
    console.log(`   ${fmtNum(t.metrics.impressions)} imp | ${fmtNum(t.metrics.likes)} likes | ${fmtNum(t.metrics.retweets)} RT | ${fmtNum(t.metrics.replies)} replies`);
    console.log(`   ${dim(t.tweet_url)}`);
    console.log();
  }
}

function printHelp(): void {
  console.log(`
Usage: xint top [options]

Show your top-performing tweets ranked by engagement metrics.

Options:
  --by <metric>     Sort by: likes, impressions, clicks, engagement_rate (default)
  --limit <N>       Number of tweets to show (default: 10)
  --since <dur>     Time period (default: 7d). e.g. 1d, 7d, 30d
  --type <type>     Filter by content type: thread, media, single
  --pages <N>       Pages of tweets to fetch (default: 3, max 10)
  --json            Output as JSON

Examples:
  xint top                           # Top 10 by engagement rate
  xint top --by likes --limit 20     # Top 20 by likes
  xint top --type media --since 30d  # Top media tweets, last 30 days
  xint top --json                    # JSON output
`);
}

File: .skills/openclaw-skills/skills/0xnyk/xint-rs/src/commands/top.rs
use anyhow::Result;
use colored::Colorize;

use crate::api::twitter;
use crate::auth::oauth;
use crate::cli::TopArgs;
use crate::client::XClient;
use crate::config::Config;
use crate::costs;
use crate::format;
use crate::models::Tweet;

pub async fn run(args: &TopArgs, config: &Config, client: &XClient) -> Result<()> {
    let client_id = config.require_client_id()?;
    let (access_token, tokens) =
        oauth::get_valid_token(client, &config.tokens_path(), client_id).await?;

    eprintln!("Fetching tweets (since {})...", args.since);

    let tweets = twitter::get_user_timeline(
        client,
        &access_token,
        &tokens.user_id,
        100,
        5,
        true,
        true,
        Some(&args.since),
    )
    .await?;

    costs::track_cost(
        &config.costs_path(),
        "timeline",
        &format!("/2/users/{}/tweets", tokens.user_id),
        tweets.len() as u64,
    );

    if tweets.is_empty() {
        println!("No tweets found in the specified time window.");
        return Ok(());
    }

    // Classify and filter by type
    let tweets: Vec<Tweet> = match args.r#type.as_deref() {
        Some("thread") => tweets
            .into_iter()
            .filter(|t| classify_type(t) == "thread")
            .collect(),
        Some("media") => tweets
            .into_iter()
            .filter(|t| classify_type(t) == "media")
            .collect(),
        Some("single") => tweets
            .into_iter()
            .filter(|t| classify_type(t) == "single")
            .collect(),
        _ => tweets,
    };

    if tweets.is_empty() {
        println!("No tweets matching the specified type filter.");
        return Ok(());
    }

    // Sort by chosen metric
    let mut sorted: Vec<&Tweet> = tweets.iter().collect();
    sorted.sort_by(|a, b| {
        let val = |t: &Tweet| -> f64 {
            match args.by.as_str() {
                "likes" => t.metrics.likes as f64,
                "impressions" => t.metrics.impressions as f64,
                "retweets" => t.metrics.retweets as f64,
                _ => engagement_rate(t),
            }
        };
        val(b)
            .partial_cmp(&val(a))
            .unwrap_or(std::cmp::Ordering::Equal)
    });

    let top: Vec<&Tweet> = sorted.into_iter().take(args.limit).collect();

    if args.json {
        let summaries: Vec<serde_json::Value> = top
            .iter()
            .enumerate()
            .map(|(i, t)| {
                serde_json::json!({
                    "rank": i + 1,
                    "id": t.id,
                    "text": t.text,
                    "type": classify_type(t),
                    "likes": t.metrics.likes,
                    "impressions": t.metrics.impressions,
                    "retweets": t.metrics.retweets,
                    "engagement_rate": (engagement_rate(t) * 100.0).round() / 100.0,
                    "url": t.tweet_url,
                    "created_at": t.created_at,
                })
            })
            .collect();
        format::print_json_pretty_filtered(&summaries)?;
    } else {
        let type_label = args.r#type.as_deref().unwrap_or("all");
        println!(
            "\n{}  Top {} tweets by {} (type: {}, since {})\n",
            "Best Performers".bold(),
            args.limit,
            args.by.cyan(),
            type_label,
            args.since
        );

        for (i, t) in top.iter().enumerate() {
            let preview = if t.text.len() > 100 {
                format!(
                    "{}...",
                    &t.text[..t
                        .text
                        .char_indices()
                        .nth(97)
                        .map(|(idx, _)| idx)
                        .unwrap_or(t.text.len())]
                )
            } else {
                t.text.clone()
            };

            let er = engagement_rate(t);
            let content_type = classify_type(t);

            println!(
                "  {}. {} [{}]\n     {} likes | {} views | {}% ER\n     {}\n",
                (i + 1).to_string().bold(),
                preview.dimmed(),
                content_type.cyan(),
                compact_number(t.metrics.likes).yellow(),
                compact_number(t.metrics.impressions),
                format!("{:.2}", er).green(),
                t.tweet_url.dimmed()
            );
        }
    }

    Ok(())
}

fn engagement_rate(t: &Tweet) -> f64 {
    if t.metrics.impressions == 0 {
        return 0.0;
    }
    let engagements = t.metrics.likes + t.metrics.retweets + t.metrics.replies + t.metrics.quotes;
    engagements as f64 / t.metrics.impressions as f64 * 100.0
}

fn classify_type(t: &Tweet) -> &'static str {
    // Thread: conversation_id matches the tweet's own id (thread starter)
    // or the tweet text is very long (note tweet / thread continuation)
    if !t.conversation_id.is_empty() && t.conversation_id == t.id && t.text.len() > 280 {
        return "thread";
    }
    // Media: has t.co links but no URL entities (typically image/video attachments)
    if t.text.contains("https://t.co/") && t.urls.is_empty() {
        return "media";
    }
    // Links: has URL entities
    if !t.urls.is_empty() {
        return "link";
    }
    "single"
}

fn compact_number(n: u64) -> String {
    if n >= 1_000_000 {
        format!("{:.1}M", n as f64 / 1_000_000.0)
    } else if n >= 1_000 {
        format!("{:.1}K", n as f64 / 1_000.0)
    } else {
        n.to_string()
    }
}
